In [174]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot  as plt 
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

# Data Understanding

### Project Dataset Overview
For this project, I used four financial datasets that collectively represent a company's profitability, growth performance, financial health, and operational efficiency. The objective was to create a Financial Performance & Corporate Benchmarking Dashboard that helps evaluate and compare companies across industries.

In [175]:
df_profit_loss_set_1 = pd.read_csv('dataset/Annual_P_L_1_final.csv')

### Annual Profit & Loss Dataset (Part 1)
What does it contain?

This dataset contains the core income statement information of companies, including:

* Sales Revenue
* Operating Profit
* EBIT
* Profit After Tax (PAT)
* Net Profit
* Earnings Per Share (EPS)
* Operating Profit Margin (OPM)
* Historical profit and revenue figures
 
#### Why is it important?

* The Profit & Loss Statement shows how effectively a company generates profits from its operations.

#### Using this dataset, we can answer questions such as:
1. Which companies generate the highest revenue?
2. Which companies are the most profitable?
3. How efficiently are companies managing their operating costs?
4. How have profits changed compared to previous years?
#### Business Value :- This dataset forms the foundation of profitability analysis and executive performance reporting.

In [176]:
df_profit_loss_set_2 = pd.read_csv('dataset/Annual_P_L_2_final.csv')

### Annual Profit & Loss Dataset (Part 2)
What does it contain?

This dataset focuses on long-term growth metrics, including:

* Sales Growth (3, 5, 7, and 10 years)
* Profit Growth (3, 5, 7, and 10 years)
* EPS Growth (3, 5, 7, and 10 years)
* Average Earnings
* Average EBIT

#### Why is it important?
* A company may have strong profits today, but analysts also want to know whether those profits have been growing consistently over time.

#### This dataset helps answer:

1. Which companies have sustainable growth?
2. Which firms consistently increase earnings?
3. Which industries are growing faster than others?
#### Business Value :- This dataset supports FP&A growth analysis and long-term performance evaluation.

In [177]:
df_bal_sheet_set = pd.read_csv('dataset/Balance_Sheet_final.csv')

### Balance Sheet Dataset
What does it contain?

This dataset represents the financial position of a company and includes:

* Total Assets
* Debt
* Equity Capital
* Reserves
* Investments
* Current Assets
* Current Liabilities
* Working Capital
* Cash Equivalents
* Inventory
* Trade Receivables

#### Why is it important?

* While the Profit & Loss Statement measures performance, the Balance Sheet measures financial strength.

#### Using this dataset, we can analyze:

1. How much debt a company carries
2. Whether it has sufficient liquidity
3. The size of its asset base
4. The strength of its capital structure
#### Business Value :- This dataset is critical for evaluating financial health, solvency, liquidity, and risk.

In [178]:
df_other_metrics_set = pd.read_csv('dataset/other_metrics_final.csv')

### Other Financial Metrics Dataset
What does it contain?

This dataset includes advanced financial and efficiency metrics such as:

* Current Ratio
* Interest Coverage Ratio
* ROCE (3-Year Average)
* CROIC
* Leverage
* Debt Capacity
* Dividend Payout Ratio
* Altman Z Score
* Enterprise Value Metrics
#### Why is it important?
These metrics help analysts move beyond raw financial statements and evaluate business quality, efficiency, and risk.

#### Using this dataset, we can answer:

1. Which companies use capital efficiently?
2. Which firms have stronger financial stability?
3. Which companies have lower financial risk?
4. Which businesses create greater value for shareholders?
#### Business Value :- This dataset enables corporate benchmarking and financial ratio analysis.

# DATA MODELING

Now we are understanding about the data but we have to make a one dataset who connect the all rest 4 of the dataset 
So we are checking that logically connect the different dataset to made a ready analysis structure 

In [179]:
datasets = {
    "P&L1": df_profit_loss_set_1,
    "P&L2": df_profit_loss_set_2,
    "Balance Sheet": df_bal_sheet_set,
    "Other Metrics": df_other_metrics_set
}
for name, df in datasets.items():

    print("\n", name)

    print("Rows:", len(df))

    print("Unique Companies:", df['Name'].nunique())

    print("Unique join_key:", df['join_key'].nunique())


 P&L1
Rows: 4668
Unique Companies: 4668
Unique join_key: 4668

 P&L2
Rows: 4668
Unique Companies: 4668
Unique join_key: 4668

 Balance Sheet
Rows: 4668
Unique Companies: 4668
Unique join_key: 4668

 Other Metrics
Rows: 4668
Unique Companies: 4668
Unique join_key: 4668


### Observation
* All four datasets contain the same number of records.
* Every row represents a unique company.
* Every company has a unique join_key.
* A company appears exactly once in every dataset.
* The dataset contains thousands of companies with one consolidated record per company.
* The four datasets have a valid one-to-one relationship through join_key and can be safely integrated after data cleaning

# Data Cleaning & Preprocessing

In [180]:
# Missing Value audit
datasets = {
    "P&L1": df_profit_loss_set_1,
    "P&L2": df_profit_loss_set_2,
    "Balance Sheet": df_bal_sheet_set,
    "Other Metrics": df_other_metrics_set
}

for name, df in datasets.items():

    print("\n" + "="*60)
    print(name)

    print("\nMissing Values (Top 20):")

    print(
        df.isnull()
          .sum()
          .sort_values(ascending=False)
          .head(20)
    )


P&L1

Missing Values (Top 20):
Net profit preceding 12months      699
Sales preceding 12months           699
OPM preceding year                 309
OPM last year                      301
OPM                                301
Material cost last year            301
NPM preceding year                 240
Return on capital employed         217
NPM last year                      208
EPS preceding year                 127
Change in promoter holding         115
Interest preceding year             43
Sales preceding year                43
Operating profit preceding year     43
Other income preceding year         43
EBIDT preceding year                43
EBIT preceding year                 43
Depreciation preceding year         43
Tax preceding year                  43
Profit after tax preceding year     43
dtype: int64

P&L2

Missing Values (Top 20):
EPS growth 10Years                   2050
Profit growth 10Years                2032
EBIDT growth 10Years                 1873
Sales growth 10Ye

## Observations
Dataset contains:
* 4,668 companies
* Most columns have less than 50% missing values
* Core financial metrics are largely available

#### Dataset 1  = profit & loss 1
1. we observe that highest missing values would be Net profit preceding 12months, Sales preceding 12months, OPM, OPM Last Year,Material Cost Last Year
2. These missing values are mostly:Historical metrics , Margin metrics, Trailing 12-month metrics
3. Important Insight Columns like: Sales, Net Profit, EBIT, Operating Profit

* Do NOT appear among the highest missing-value columns So, Core profitability analysis is still reliable

#### Dataset 2 = Profit & loss 2
1. Largest missing values: EPS Growth 10Y, Profit Growth 10Y, Sales Growth 10Y 
* These missing values are likely due to: Newer companies, Recently listed companies, Insufficient historical records

#### Dataset 3 - Balance Sheet
1. Most missing values occur in: Debt 10 Years Back, Working Capital 10 Years Back, Debt 7 Years Back
* Again, the missing values are concentrated in historical trend fields, not current financial health fields.

#### Dataset 4 - Other metrics 
1. Highest missing values : PEG Ratio, Intrinsic Value, PB X PE, Graham
2. These are derived valuation metrics they often become unavailable when: Earnings are negative, Growth is negative,Inputs required for calculation are missing.
* Missing valuation metrics do not automatically indicate bad data quality many times they indicate: Metrics cannot be calculated

## Insights
The missing values are NOT randomly distributed

They are concentrated in:
* Category 1 : Long-term historical fields
* Category 2 : Advanced valuation metrics
* Category 3 : Some ratio calculations
## Conclusion
The dataset has manageable missing values, primarily concentrated in long-term historical and advanced valuation metrics. Core financial statement and profitability variables remain largely complete, making the dataset suitable for financial performance analysis and corporate benchmarking.

In [181]:
## Datatype Audit
datasets = {
    "P&L1": df_profit_loss_set_1,
    "P&L2": df_profit_loss_set_2,
    "Balance Sheet": df_bal_sheet_set,
    "Other Metrics": df_other_metrics_set
}

for name, df in datasets.items():

    print("\n" + "="*60)
    print(name)

    print(df.dtypes.value_counts())


P&L1
float64    54
object      4
Name: count, dtype: int64

P&L2
float64    26
object      4
Name: count, dtype: int64

Balance Sheet
float64    47
object      4
Name: count, dtype: int64

Other Metrics
float64    40
object      4
Name: count, dtype: int64


## Observation 

1. Usually when analysts receive raw financial data: Revenue comes as text, Percentages come as strings, Currency values contain commas, Ratios are stored as objects
* Most columns are already numeric.1. This is actually a very clean financial dataset.

2. We still don't know what those 4 object columns are : Name, BSE Code, NSE Code, Industry, join_key
* To check this columns are object or not we run a function 

In [182]:
for name, df in datasets.items():

    print("\n" + "="*50)
    print(name)

    print(df.select_dtypes(include='object').columns.tolist())


P&L1
['Name', 'NSE Code', 'Industry', 'join_key']

P&L2
['Name', 'NSE Code', 'Industry', 'join_key']

Balance Sheet
['Name', 'NSE Code', 'Industry', 'join_key']

Other Metrics
['Name', 'NSE Code', 'Industry', 'join_key']


## Observation
* There are no financial metrics stored as text.
* Object columns are exactly what we expect.
* join_key is consistently stored as object in all datasets.

In [183]:
#Duplicate audit
for name, df in datasets.items():
    print(name, "=", df.duplicated().sum())

P&L1 = 0
P&L2 = 0
Balance Sheet = 0
Other Metrics = 0


# Observation
* No duplicate records exist inside the dataset.

## Data Quality Assessment
We audited: Structure, Relationships, Datatypes, Missing values, Duplicates
* conclusion : The dataset is structurally strong and analysis-ready, with missing values being the primary data quality consideration
* conclusion : The financial datasets contain 4,668 unique companies with no duplicate records and a consistent one-to-one relationship across all source tables. Numeric financial metrics are correctly stored in numerical formats, and identifier fields are consistently represented. The primary data quality challenge relates to missing values in historical growth and valuation metrics rather than structural data issues

In [184]:
for name, df in datasets.items():
    missing_pct = (df.isnull().sum() / len(df) * 100).sort_values(ascending=False)

    print("\n" + "="*60)
    print(name)

    print(missing_pct[missing_pct > 0].head(20))


P&L1
Net profit preceding 12months      14.974293
Sales preceding 12months           14.974293
OPM preceding year                  6.619537
OPM last year                       6.448158
OPM                                 6.448158
Material cost last year             6.448158
NPM preceding year                  5.141388
Return on capital employed          4.648672
NPM last year                       4.455870
EPS preceding year                  2.720651
Change in promoter holding          2.463582
Interest preceding year             0.921165
Sales preceding year                0.921165
Operating profit preceding year     0.921165
Other income preceding year         0.921165
EBIDT preceding year                0.921165
EBIT preceding year                 0.921165
Depreciation preceding year         0.921165
Tax preceding year                  0.921165
Profit after tax preceding year     0.921165
dtype: float64

P&L2
EPS growth 10Years                   43.916024
Profit growth 10Years     

## Missing Value Classification
Low Missing (< 10%) :- Usually manageable.
action :- Investigate, Recalculate if possible, Otherwise preserve

Medium Missing (10%–30%) :- Need business justification.
action :- Column-by-column decision

High Missing (> 30%) :- Very careful.
action :- Usually preserve as NaN, Don't force imputation,Evaluate business importance

In [185]:
missing_reports = {}

for name, df in datasets.items():

    report = pd.DataFrame({
        'missing_count': df.isnull().sum(),
        'missing_percentage': round((df.isnull().sum()/len(df))*100, 2)
    })

    report = (
        report
        .query("missing_count > 0")
        .sort_values("missing_percentage", ascending=False)
    )

    missing_reports[name] = report

In [186]:
missing_reports['P&L1'].head(20)

,missing_count,missing_percentage
Net profit preceding 12months,699,14.97
Sales preceding 12months,699,14.97
OPM preceding year,309,6.62
OPM,301,6.45
Material cost last year,301,6.45
OPM last year,301,6.45
NPM preceding year,240,5.14
Return on capital employed,217,4.65
NPM last year,208,4.46
EPS preceding year,127,2.72


#### df_profit_loss_set_1 :- This dataset is extremely healthy most important profitability metrics are nearly complete.
Decision:
* Preserve historical missing values.
* Investigate OPM reconstruction.
* No column dropping.

In [187]:
missing_reports['P&L2'].head(20)

,missing_count,missing_percentage
EPS growth 10Years,2050,43.92
Profit growth 10Years,2032,43.53
EBIDT growth 10Years,1873,40.12
Sales growth 10Years,1732,37.10
Profit growth 7Years,1617,34.64
EPS growth 7Years,1617,34.64
EBIDT growth 7Years,1431,30.66
Sales growth 7Years,1338,28.66
EPS growth 5Years,1241,26.59
Average Earnings 10Year,1234,26.44


#### df_profit_loss_set_2 :- This is where most missing values exist.
Most missing values belong :- 5-year growth, 7-year growth, 10-year growth

Decision:
* A company cannot have: 10-year growth, 7-year growth
* Keep NaN

In [188]:
missing_reports['Balance Sheet'].head(20)

,missing_count,missing_percentage
Debt 10Years back,1336,28.62
Working capital 10Years back,1336,28.62
Net block 7Years back,945,20.24
Working capital 7Years back,945,20.24
Debt 7Years back,945,20.24
Working capital 5Years back,581,12.45
Net block 5Years back,581,12.45
Debt 5Years back,581,12.45
Working capital 3Years back,202,4.33
Debt 3Years back,202,4.33


#### df_bal_sheet_set :

Highest missing:- Debt 10Y Back, Working Capital 10Y Back, Debt 7Y Back

Interpretation :-Historical balance-sheet history is unavailable for some firms.

Decision:
* Keep NaN.

In [189]:
missing_reports['Other Metrics'].head(20)

,missing_count,missing_percentage
PEG Ratio,1575,33.74
Intrinsic Value,1269,27.19
PB X PE,1230,26.35
Graham,1224,26.22
Interest Coverage Ratio,761,16.30
Interest Coverage,761,16.30
QoQ Sales,544,11.65
Market cap to quarterly profit,522,11.18
Mkt Cap To Debt Cap,425,9.10
Debt Capacity,421,9.02


#### df_other_metrics_set :

High missing:- PEG Ratio, Intrinsic Value, Graham

These are derived valuation metrics missing often occurs because Negative earnings, Missing growth, Calculation not possible

Decision:
* Keep NaN.


### Observations :
1. The dataset exhibits relatively low missingness in core financial variables and moderate missingness in selected historical metrics.
2. Missing values appear to be driven by limited historical reporting availability rather than data extraction errors.
3. Since the missing values are concentrated in historical growth metrics rather than current profitability indicators, the dataset remains suitable for financial performance analysis and benchmarking.

### Missing Value Decision

* After performing a detailed missing value assessment, it was observed that most missing values were concentrated in long-term historical metrics (such as 5-year, 7-year, and 10-year growth indicators), historical balance sheet fields, and advanced valuation metrics (such as PEG Ratio, Graham Value, and Intrinsic Value). These missing values primarily represent the unavailability of historical or calculable financial information rather than data quality issues.

* Therefore, no blanket imputation technique (such as filling with 0, mean, or median) was applied. The missing values were retained as NaN to preserve their original business meaning and avoid introducing misleading financial information into the analysis. Any future treatment will be considered only for metrics that can be reliably reconstructed using available financial data and valid financial formulas.

# Column Standardization & Dataset Preparation

#### Now we perform the column standardization to improve data usability, ensure cross-tool compatibility, and establish a consistent naming convention before data integration and analysis.

In [190]:
df_profit_loss_set_1.columns = (df_profit_loss_set_1.columns.str.lower().str.replace(" ", "_", regex=False))
df_profit_loss_set_2.columns = (df_profit_loss_set_2.columns.str.lower().str.replace(" ", "_", regex=False))
df_bal_sheet_set.columns = (df_bal_sheet_set.columns.str.lower().str.replace(" ", "_", regex=False))
df_other_metrics_set.columns = (df_other_metrics_set.columns.str.lower().str.replace(" ", "_", regex=False))

In [191]:
print(df_profit_loss_set_1.columns[:10])
print(df_profit_loss_set_2.columns[:10])
print(df_bal_sheet_set.columns[:10])
print(df_other_metrics_set.columns[:10])

Index(['name', 'bse_code', 'nse_code', 'industry', 'current_price', 'sales',
       'opm', 'profit_after_tax', 'return_on_capital_employed', 'eps'],
      dtype='object')
Index(['name', 'bse_code', 'nse_code', 'industry', 'current_price',
       'sales_growth_3years', 'sales_growth_5years', 'profit_growth_3years',
       'profit_growth_5years', 'sales_growth_10years_median'],
      dtype='object')
Index(['name', 'bse_code', 'nse_code', 'industry', 'current_price', 'debt',
       'equity_capital', 'preference_capital', 'reserves', 'secured_loan'],
      dtype='object')
Index(['name', 'bse_code', 'nse_code', 'industry', 'current_price',
       'price_to_sales', 'price_to_free_cash_flow', 'evebitda',
       'current_ratio', 'interest_coverage_ratio'],
      dtype='object')


Action performed :
* Converted all column names to lowercase.
* Replaced spaces with underscores (_).
* Established a uniform naming convention across all financial datasets.

## Merge Readiness Check

In [192]:
for name, df in datasets.items():
    print(f"\n{name}")
    print(df['join_key'].dtype)


P&L1
object

P&L2
object

Balance Sheet
object

Other Metrics
object


* The common identifier (join_key) is consistently stored as an object datatype across all datasets, confirming that the datasets are technically ready for integration from a key-format perspective.

In [193]:
common_cols = (
    set(df_profit_loss_set_1.columns)
    & set(df_profit_loss_set_2.columns)
    & set(df_bal_sheet_set.columns)
    & set(df_other_metrics_set.columns)
)

print(sorted(common_cols))

['bse_code', 'current_price', 'industry', 'join_key', 'market_capitalization', 'name', 'nse_code']


* Before merging we check that which columns are identifiers and should appear only once in the final dataset

In [194]:
print(set(df_profit_loss_set_1['join_key']) == set(df_profit_loss_set_2['join_key']))

print(set(df_profit_loss_set_1['join_key']) == set(df_bal_sheet_set['join_key']))

print(set(df_profit_loss_set_1['join_key']) == set(df_other_metrics_set['join_key']))

True
True
True


#### Objective:
* Before integrating the financial datasets, a merge readiness assessment was performed to validate the consistency and compatibility of the source tables.

#### Validation Checks Performed:
* Verified that all datasets contained the same number of company records.
* Confirmed that each dataset had a unique join_key for every company.
* Checked that the join_key datatype was consistent across all datasets.
* Identified common columns shared across multiple datasets.
* Evaluated potential duplicate columns that could create redundancy after merging.

By performing a merge readiness assessment, the integrity of the data model was validated before integration.

In [195]:
duplicate_cols = ['name','bse_code','nse_code','industry','current_price','market_capitalization']
print(duplicate_cols)

['name', 'bse_code', 'nse_code', 'industry', 'current_price', 'market_capitalization']


* Common descriptive attributes such as company name, industry, market capitalization, and stock identifiers were retained only once in the base table to avoid redundancy. The join_key was preserved because it served as the primary integration key across all financial datasets

In [196]:
pnl2_merge = df_profit_loss_set_2.drop(columns=duplicate_cols)
balance_merge = df_bal_sheet_set.drop(columns=duplicate_cols)
metrics_merge = df_other_metrics_set.drop(columns=duplicate_cols)

In [197]:
print(df_profit_loss_set_1.shape)
print(pnl2_merge.shape)
print(balance_merge.shape)
print(metrics_merge.shape)

(4668, 58)
(4668, 24)
(4668, 45)
(4668, 38)


* Prior to integration, duplicate company-level attributes were removed from the secondary datasets. This ensured that each dataset contributed only unique financial information, preventing redundant columns and improving the quality of the final analytical model.

In [198]:
master_df = df_profit_loss_set_1.copy()

In [199]:
master_df = master_df.merge(pnl2_merge,on='join_key',how='left')
master_df = master_df.merge(balance_merge,on='join_key',how='left')
master_df = master_df.merge(metrics_merge,on='join_key',how='left')

In [200]:
print(master_df.shape)

(4668, 162)


In [201]:
master_df['join_key'].nunique()

4668

In [202]:
master_df.head()

,name,bse_code,nse_code,industry,current_price,sales,opm,profit_after_tax,return_on_capital_employed,eps,...,mkt_cap_to_debt_cap,dividend_payout_ratio,graham,price_to_cash_flow,roce3yr_avg,pb_x_pe,ncavps,market_capt_to_cash_flow,altman_z_score,market_cap_to_quarterly_profit
0,20 Microns,533022.0,20MICRONS,Mining / Minerals / Metals,224.55,777.49,13.59,57.38,21.70,15.89,...,1.27,7.87,189.11,14.58,20.14,31.58,70.17,14.58,5.36,58.18
1,21st Cent. Mgmt.,526921.0,21STCENMGM,Finance & Investments,70.44,34.92,94.27,32.23,73.00,30.70,...,-3.28,0.00,187.38,-42.02,-1.17,3.20,38.27,-42.02,6.44,4.36
2,360 ONE,542772.0,360ONE,Finance & Investments,1009.40,2920.91,58.37,804.18,14.47,22.41,...,11.77,73.64,220.16,-76.81,13.73,464.13,-9.08,-76.80,4.67,148.82
3,3B Blackbio,532067.0,__NA__,Healthcare,1182.00,74.12,44.50,32.10,20.95,38.13,...,2.09,0.00,472.22,37.34,22.23,140.40,123.17,37.26,9.14,114.24
4,3C IT Solutions,544190.0,__NA__,Computers - Software - Medium / Small,44.01,61.93,5.17,1.14,25.63,31.67,...,NaN,0.00,NaN,-12.27,27.48,NaN,28.35,-12.31,5.00,NaN


### Phase 3: Master Dataset Creation and Validation

A consolidated master financial dataset was created by integrating Profit & Loss, Growth, Balance Sheet, and Financial Ratio datasets using a one-to-one relationship based on the `join_key` field.

The Profit & Loss dataset was selected as the base table, and additional datasets were merged using a left join strategy to preserve the complete company universe while enriching each record with growth, financial health, valuation, and benchmarking metrics.

Post-merge validation confirmed that the dataset maintained data integrity throughout the integration process:

* Total Rows: 4,668
* Unique Companies: 4,668
* Unique Join Keys: 4,668
* Duplicate Columns: None
* Missing Join Keys: 0
* Total Features: 162

These validation checks confirmed that the master dataset retained a one-company-per-record structure without introducing duplicate entities or data inconsistencies.

With a clean and validated analytical dataset established, the project is now ready for Exploratory Data Analysis (EDA), KPI development, financial benchmarking, SQL analysis, and Power BI dashboard creation.


# Exploratory Data Analysis (EDA)

After creating the master financial dataset, I retained all 162 columns because each metric could provide value in later stages of the project, such as KPI development, benchmarking, SQL analysis, and Power BI reporting.

However, during Exploratory Data Analysis (EDA), I did not analyze all 162 columns simultaneously.

As a Financial Analyst, my objective was to answer specific business questions rather than perform random analysis on every available metric.

Therefore, I grouped the dataset into business-focused analytical categories:

* Profitability Analysis
* Growth Analysis
* Financial Health Analysis
* Valuation Analysis
* Corporate Benchmarking

For each analysis area, only the relevant metrics were selected.

The remaining columns were retained in the master dataset but were not required for that specific business objective.


In [203]:
master_df.columns.tolist()

['name',
 'bse_code',
 'nse_code',
 'industry',
 'current_price',
 'sales',
 'opm',
 'profit_after_tax',
 'return_on_capital_employed',
 'eps',
 'change_in_promoter_holding',
 'sales_last_year',
 'operating_profit_last_year',
 'other_income_last_year',
 'ebidt_last_year',
 'depreciation_last_year',
 'ebit_last_year',
 'interest_last_year',
 'profit_before_tax_last_year',
 'tax_last_year',
 'profit_after_tax_last_year',
 'extraordinary_items_last_year',
 'net_profit_last_year',
 'dividend_last_year',
 'material_cost_last_year',
 'employee_cost_last_year',
 'opm_last_year',
 'npm_last_year',
 'operating_profit',
 'interest',
 'depreciation',
 'eps_last_year',
 'ebit',
 'net_profit',
 'current_tax',
 'tax',
 'other_income',
 'last_annual_result_date',
 'sales_preceding_year',
 'operating_profit_preceding_year',
 'other_income_preceding_year',
 'ebidt_preceding_year',
 'depreciation_preceding_year',
 'ebit_preceding_year',
 'interest_preceding_year',
 'profit_before_tax_preceding_year',
 '

## PROFITABILITY ANALYSIS

Profitability tells us:
`How much money a company makes`

In [204]:
top_sales = (master_df[['name','industry','sales']].sort_values(by='sales',ascending=False).head(10))
top_sales

,name,industry,sales
3308,Reliance Industr,Refineries,901064.00
2283,Life Insurance,Finance & Investments,846656.19
1684,I O C L,Refineries,776351.85
2805,O N G C,Oil Drilling / Allied Services,643037.01
452,B P C L,Refineries,448083.03
3893,St Bk of India,Banks - Public Sector,439188.51
4114,Tata Motors,Automobiles - LCVs / HCVs,437927.77
1544,H P C L,Refineries,433856.51
1588,HDFC Bank,Banks - Private Sector,283649.02
3219,Rajesh Exports,Diamond Cutting / Jewellery,280676.35


In [205]:
top_pat = (master_df[['name','industry','profit_after_tax']].sort_values(by='profit_after_tax',ascending=False).head(10))
top_pat

,name,industry,profit_after_tax
3308,Reliance Industr,Refineries,69621.00
3893,St Bk of India,Banks - Public Sector,67102.23
1588,HDFC Bank,Banks - Private Sector,64062.04
2805,O N G C,Oil Drilling / Allied Services,50269.36
4135,TCS,Computers - Software - Large,46598.19
1693,ICICI Bank,Banks - Private Sector,44256.37
1684,I O C L,Refineries,41729.69
2283,Life Insurance,Finance & Investments,40884.77
813,Coal India,Mining / Minerals / Metals,37369.13
4114,Tata Motors,Automobiles - LCVs / HCVs,32193.28


In [206]:
bottom_pat = (master_df[['name','industry','profit_after_tax']].sort_values(by='profit_after_tax',ascending=True).head(10))
bottom_pat

,name,industry,profit_after_tax
4535,Vodafone Idea,Telecommunications - Service Provider,-31973.31
1290,Future Retail,Trading,-4445.43
4327,Unitech,Construction,-3326.65
2335,M T N L,Telecommunications - Service Provider,-3267.52
1027,Dish TV India,Entertainment / Electronic Media Software,-3161.81
1403,Glenmark Pharma.,Pharmaceuticals - Indian - Bulk Drugs & Formln,-2797.24
3310,Reliance Power,Power Generation And Supply,-2197.91
3272,RattanIndia Pow.,Power Generation And Supply,-1738.33
2777,NMDC Steel,Steel - Medium / Small,-1560.32
1288,Future Lifestyle,Trading,-1500.62


In [207]:
industry_profit = (master_df.groupby('industry')[['sales','profit_after_tax']].sum().sort_values(by='profit_after_tax',ascending=False))
industry_profit.head(15)

,sales,profit_after_tax
industry,,
Banks - Private Sector,888319.59,193842.40
Refineries,2720306.22,160882.96
Finance & Investments,1944029.63,156595.69
Banks - Public Sector,1101191.51,150638.95
Computers - Software - Large,713527.81,109972.04
Power Generation And Supply,467919.48,73125.12
Mining / Minerals / Metals,380185.05,58652.58
Oil Drilling / Allied Services,678296.64,57237.53
Miscellaneous,522001.17,42464.96


` Finding 1: Revenue Concentration is High`
* Revenue generation is highly concentrated among a small group of large-cap companies. Energy, Refining, Financial Services, and Banking sectors dominate the revenue landscape, indicating significant scale advantages and market leadership.

`Finding 2: Profitability is Diversified Across Industries`
* Unlike revenue concentration, profitability is distributed across multiple sectors including Energy, Banking, Oil & Gas, and Technology, suggesting diversified profit creation within the Indian corporate ecosystem.

`Finding 3: Private Banking is the Profitability Champion`
* Private sector banks emerged as the most profitable industry group, generating the highest aggregate profit after tax. Strong lending franchises, higher operational efficiency, and better asset quality may contribute to superior profitability.

`Finding 4: Refineries Dominate Revenue but Not Profitability`
* Although the Refinery sector generated the highest revenue in the dataset, Private Sector Banks delivered higher aggregate profitability. This highlights that scale alone does not guarantee superior earnings performance.

`Finding 5: Software Industry Shows Exceptional Efficiency`
* Large software companies generated substantial profitability despite lower revenue compared with Banking and Refining sectors, indicating superior margin profiles and asset-light business models.

`Finding 6: Telecommunications Appears Financially Stressed`
The telecommunications sector displayed significant profitability challenges, with multiple companies reporting substantial losses. Competitive pressures, high infrastructure costs, and leverage burdens may be impacting sector performance.

`Finding 7: Power Sector is Mixed`
* While the Power Generation sector contributes meaningfully to overall profitability, company-level performance varies considerably, suggesting uneven operational efficiency across industry participants.

## Growth Analysis

Growth tells us: `Which companies are growing fastest?`

In [208]:
growth_cols = ['sales_growth_3years','sales_growth_5years','profit_growth_3years','profit_growth_5years','eps_growth_3years','eps_growth_5years']
master_df[growth_cols].describe()

,sales_growth_3years,sales_growth_5years,profit_growth_3years,profit_growth_5years,eps_growth_3years,eps_growth_5years
count,4055.000000,3691.000000,3891.000000,3453.000000,3839.000000,3427.000000
mean,31.067598,11.617242,49.636137,23.728283,33.777593,18.354129
std,107.614626,38.630609,96.934858,40.765630,75.239542,37.478854
min,-95.090000,-83.640000,-80.940000,-66.030000,-100.000000,-78.480000
25%,4.785000,-0.730000,6.905000,4.560000,0.495000,1.380000
50%,17.240000,7.960000,27.870000,16.260000,23.340000,14.660000
75%,32.595000,16.845000,60.130000,32.360000,47.270000,27.920000
max,3728.060000,659.380000,2022.560000,780.360000,1582.720000,681.940000


Observation
1. Sales Growth vs Profit Growth:- Companies are growing profits faster than revenues, suggesting improving operating efficiency, better cost management, and stronger margin expansion across a significant portion of the dataset.

2. EPS Growth:- Earnings growth is broadly aligned with profit growth, indicating that profitability improvements are translating into shareholder value rather than being diluted through excessive equity issuance.

3. Negative Growth Exists:- While many companies exhibited healthy growth, the dataset also contains firms experiencing severe revenue and earnings contraction, highlighting uneven business performance across sectors.

4. Extreme Growth Outliers:- A small subset of companies reported exceptionally high growth rates, likely driven by turnaround situations, low historical bases, or extraordinary business events. Such outliers should be interpreted cautiously when benchmarking performance.

A company growing revenue above 17%, profit above 28%, and EPS above 23% can be considered above-average relative to the dataset.

In [209]:
# Top Sales Growth Companies\
master_df[['name','industry','sales_growth_3years']].sort_values(by='sales_growth_3years',ascending=False).head(10)

,name,industry,sales_growth_3years
3271,RattanIndia Ent,Miscellaneous,3728.06
1414,Globe Commercial,Trading,2432.17
3023,PNGS Gargi FJ,Trading,1615.43
4138,Teamo Production,Finance & Investments,1560.64
246,Andhra Cements,Cement - South India,1289.22
3321,Resgen,Refineries,1215.21
2295,Lloyds Enterpris,Trading,962.06
2557,Modern Engineer.,Construction,952.30
3111,Prithvi Exchange,Finance & Investments,852.75
2153,KKRRAFTON Develo,Computers - Hardware,850.48


In [210]:
# Top Profit Growth Companies
master_df[['name','industry','profit_growth_3years']].sort_values(by='profit_growth_3years',ascending=False).head(10)

,name,industry,profit_growth_3years
2297,Lloyds Metals,Mining / Minerals / Metals,2022.56
983,Dhampur Bio,Sugar,1442.77
1386,GHCL Textiles,Textiles - Spinning - Synthetic / Blended,1246.63
2105,Kesar India,Construction,902.33
841,Confidence Fut.,Trading,847.27
727,Cellecor Gadgets,Trading,830.83
2295,Lloyds Enterpris,Trading,779.46
3321,Resgen,Refineries,769.34
94,Addictive Learn,Miscellaneous,768.45
4239,Transteel Seatin,Miscellaneous,726.70


In [211]:
#Top EPS Growth Companies
master_df[['name','industry','eps_growth_3years']].sort_values(by='eps_growth_3years',ascending=False).head(10)

,name,industry,eps_growth_3years
2297,Lloyds Metals,Mining / Minerals / Metals,1582.72
2295,Lloyds Enterpris,Trading,747.77
2287,Linc,Printing & Stationery,686.92
4527,Vivid Mercantile,Trading,657.50
3271,RattanIndia Ent,Miscellaneous,619.94
3050,Porwal Auto Comp,Castings & Forgings,610.72
4200,Tinna Rubber,Miscellaneous,596.01
1414,Globe Commercial,Trading,574.53
4053,Swiss Military,Trading,549.51
1272,Franklin Indust.,Trading,540.06


Observation 1: Growth Leaders ≠ Large Companies

The fastest-growing companies are primarily small and mid-sized firms rather than large-cap market leaders. This suggests that growth opportunities are often concentrated among emerging businesses with lower revenue bases.

Observation 2: Small Base Effect

Extremely high growth rates should be interpreted cautiously, as they may be driven by small historical revenue bases rather than large-scale business expansion.

Observation 3: Trading Industry Appears Repeatedly

Trading-related companies appear frequently among the highest-growth firms, indicating strong momentum within selected segments of the trading and distribution ecosystem.

Observation 4: Lloyds Enterprise is a Multi-Metric Growth Leader

Lloyds Enterprise demonstrated balanced growth across revenue, profitability, and earnings per share, making it one of the strongest overall growth performers in the dataset.

Observation 5: Lloyds Metals is the Profitability Growth Champion :-
Lloyds Metals delivered exceptional profit and EPS growth, indicating that earnings expansion translated directly into shareholder value creation.

Observation 6: Revenue Growth vs Profit Growth :-
Revenue expansion alone is not sufficient to create value. Sustainable growth requires the conversion of revenue gains into profitability and earnings growth.

* Growth across the dataset is highly uneven, with a small group of companies delivering exceptionally strong revenue, profit, and earnings growth. Trading, Mining, and selected Industrial businesses appear prominently among the top growth performers. While several firms reported extraordinary growth rates, many of these cases may reflect low-base effects and should be evaluated alongside company size and profitability metrics. Companies such as Lloyds Enterprise and Lloyds Metals demonstrated balanced growth across multiple financial dimensions, making them standout performers within the dataset.

In [212]:
industry_growth = (master_df.groupby('industry')[['sales_growth_3years','profit_growth_3years','eps_growth_3years']].mean().sort_values(
        by='sales_growth_3years',
        ascending=False
    )
)

industry_growth.head(15)

,sales_growth_3years,profit_growth_3years,eps_growth_3years
industry,,,
Refineries,178.073750,134.945000,40.892500
Cement - South India,109.123571,-3.018889,-7.116667
Transmisson Line Towers / Equipment,102.675714,121.878889,0.700000
Recreation / Amusement Parks,93.830000,72.698000,72.222000
Travel Agencies,83.198000,71.214444,34.141000
Computers - Hardware,74.466923,89.015455,-9.524545
Moulded Luggage,60.755000,77.690000,75.200000
Domestic Appliances,57.744286,26.288571,-2.742857
Computers - Education,57.420000,89.602222,62.894444


*  Refineries:- The Refinery sector emerged as one of the strongest performing industries across both profitability and growth dimensions. High revenue growth combined with substantial profit growth suggests favorable industry conditions and strong operational performance.
* Transmission Line Towers / Equipment:- Profit growth exceeded revenue growth, indicating improved operating efficiency and margin expansion within the sector.
* Travel Agencies:- Travel-related businesses demonstrated strong recovery-led growth, reflecting improving demand conditions and sector normalization.
* Trading Sector Appears Again:- Trading businesses exhibited strong growth momentum at both the company and industry level, indicating broad-based expansion across the sector.
* Finance & Investments:- Finance and Investment companies demonstrated balanced growth across revenue, profitability, and shareholder earnings, reflecting sustainable business expansion.

`Despite strong revenue growth, profitability and earnings declined, suggesting cost pressures and margin compression within the sector.`

## Financial Health Analysis

Financial Health tells us:`Whether the company can survive and sustain growth`

In [213]:
health_cols = ['debt','equity_capital','working_capital','current_ratio','interest_coverage_ratio','leverage','net_worth','debt_capacity',
               'total_assets']
master_df[health_cols].describe()

,debt,equity_capital,working_capital,current_ratio,interest_coverage_ratio,leverage,net_worth,debt_capacity,total_assets
count,4.652000e+03,4632.000000,4652.000000,4629.000000,3907.000000,4606.000000,4632.000000,4247.000000,4.652000e+03
mean,6.557444e+03,148.928025,1008.062814,12.190073,52.412009,1.935532,2227.871965,41.791879,1.189545e+04
std,1.044876e+05,1268.927764,12099.277970,107.214725,666.366275,18.635462,17726.354452,2755.509466,1.508403e+05
min,0.000000e+00,0.000000,-108607.390000,-1156.000000,-3401.000000,-1099.490000,-104166.800000,-166.820000,0.000000e+00
25%,1.607500e+00,6.017500,6.622500,1.190000,1.440000,1.120000,14.525000,-0.210000,3.499000e+01
50%,2.103000e+01,12.815000,42.910000,1.880000,3.900000,1.520000,68.945000,0.010000,1.562300e+02
75%,1.469750e+02,30.000000,229.752500,3.770000,13.995000,2.380000,511.777500,0.160000,1.045662e+03
max,5.606147e+06,50119.800000,475932.030000,3431.000000,34553.000000,335.270000,793481.000000,179572.780000,6.733756e+06


Observation :
* Debt Distribution is Highly Skewed :- A few very large companies carry massive debt loads, pulling the average upward.
* Liquidity Position Looks Healthy :- Most companies appear capable of meeting short-term obligations since a Current Ratio above 1 is generally considered acceptable.
* Interest Coverage Looks Comfortable :- A typical company generates nearly 4 times its interest expense, suggesting manageable debt servicing ability.
* Leverage Appears Moderate :- Most companies are not excessively debt-dependent, although some highly leveraged outliers exist.
* Working Capital Distribution is Uneven :- Large firms are significantly increasing the average. Working capital strength is concentrated among bigger companies.

In [214]:
#Top Financially Strong Companies
master_df[['name','industry','current_ratio','interest_coverage_ratio','net_worth']].sort_values(by='net_worth',ascending=False).head(10)

,name,industry,current_ratio,interest_coverage_ratio,net_worth
3308,Reliance Industr,Refineries,1.18,5.53,793481.00
3893,St Bk of India,Banks - Public Sector,2.56,1.35,414939.17
2805,O N G C,Oil Drilling / Allied Services,0.88,8.70,337070.23
1588,HDFC Bank,Banks - Private Sector,3.38,1.50,289437.50
1693,ICICI Bank,Banks - Private Sector,2.20,1.83,214497.79
1684,I O C L,Refineries,0.73,8.32,183416.27
2795,NTPC,Power Generation And Supply,0.85,3.34,160709.27
1967,Jio Financial,Finance & Investments,98.41,191.45,139147.66
440,Axis Bank,Banks - Private Sector,2.84,1.57,129781.71
497,Bank of Baroda,Banks - Public Sector,1.80,1.37,119712.24


In [215]:
# High Debt company 
master_df[['name','industry','debt']].sort_values(by='debt', ascending=False).head(10)

,name,industry,debt
3893,St Bk of India,Banks - Public Sector,5606146.99
1588,HDFC Bank,Banks - Private Sector,2139211.91
497,Bank of Baroda,Banks - Public Sector,1453760.94
3143,Punjab Natl.Bank,Banks - Public Sector,1451810.86
1693,ICICI Bank,Banks - Private Sector,1399893.96
687,Canara Bank,Banks - Public Sector,1369780.03
4317,Union Bank (I),Banks - Public Sector,1163058.51
440,Axis Bank,Banks - Private Sector,1152038.29
3052,Power Fin.Corpn.,Finance - Term-Lending Institutions,861961.06
498,Bank of India,Banks - Public Sector,821571.71


In [216]:
#Most leveraged company
master_df[['name','industry','leverage']].sort_values(by='leverage', ascending=False).head(10)

,name,industry,leverage
2129,Kimia Bioscien.,Pharmaceuticals - Indian - Bulk Drugs,335.27
1014,Digicontent,Entertainment / Electronic Media Software,202.30
329,Artson Engg.,Engineering - Turnkey Services,166.92
2938,Parmax Pharma,Pharmaceuticals - Indian - Bulk Drugs,149.73
3927,Sterling Powerg.,Electric Equipment,119.68
2892,Pacheli Indust.,Finance & Investments,111.43
3441,Sagar Diamonds,Diamond Cutting / Jewellery,100.64
3453,Sai Swami Metals,Domestic Appliances,73.83
3962,Suncity Synth.,Textiles - Processing,68.31
2283,Life Insurance,Finance & Investments,59.79


Observation 
* `Banking Sector Dominates Debt Rankings`:- The highest debt levels are concentrated within banking and financial institutions. This is expected because borrowing and lending activities form the core operating model of financial service companies. Therefore, high debt levels in these sectors should not automatically be interpreted as financial distress.
* `Debt Must Be Viewed in Industry Context`:- Debt analysis should always be performed relative to industry structure. Financial institutions naturally operate with significantly higher leverage than non-financial businesses.
* `Extreme Leverage Risk Exists in Selected Companies`:- Several companies exhibit exceptionally high leverage ratios, indicating elevated dependence on debt financing and potentially higher financial risk during adverse business conditions.
* `Public Sector Banks Hold Strong Net Worth`:- Large-cap corporations and leading financial institutions maintain substantial net worth positions, providing strong capital bases and financial resilience.

In [217]:
industry_health = (master_df.groupby('industry')[['current_ratio','interest_coverage_ratio','leverage','net_worth']].mean())
industry_health.head(15)

,current_ratio,interest_coverage_ratio,leverage,net_worth
industry,,,,
Abrasives And Grinding Wheels,3.577500,54.920000,1.217500,1358.350000
Air-conditioners,1.270000,4.802500,2.245000,2476.437500
Aluminium and Aluminium Products,1.926667,11.667647,1.821176,6778.559444
Aquaculture,8.940000,0.848750,5.645455,105.319091
Auto Ancillaries,-8.849727,121.009074,1.743962,1199.087037
Automobiles - LCVs / HCVs,1.222000,5.004000,3.658000,19475.370000
Automobiles - Motorcycles / Mopeds,1.256000,37.134000,2.476000,8560.908000
Automobiles - Passenger Cars,1.895000,238.222500,2.087500,37980.892500
Automobiles - Scooters And 3 - Wheelers,1.046667,56.563333,-0.276667,9786.663333


 * `Public & Private Banks Have Strong Capital Base` :- Both public and private sector banks maintain significant net worth levels, supporting their ability to absorb risk and sustain lending operations.

 * `Private & Public Banks Operate With Higher Leverage` :- Banking institutions exhibit structurally higher leverage ratios compared to most industries due to the nature of deposit-based financing and credit operations.

* `Automobile Industry Shows Balanced Financial Health` :-Automobile manufacturers demonstrate healthy liquidity positions and strong debt-servicing capacity, indicating robust financial stability.

* `High Current Ratio Can Be a Warning Too` :- Extremely high current ratios may indicate underutilized assets or inefficient working capital deployment rather than superior financial performance.

## Executive Summary

1. Which Industries Are Most Profitable?

The profitability analysis revealed that Refineries, Private Sector Banks, Finance & Investments, and Public Sector Banks emerged as the most profitable industries in the dataset.

2. Which Industries Are Growing Fastest?

Industry growth benchmarking identified Refineries, Transmission Infrastructure, Travel Services, Trading, and Finance & Investments as high-growth sectors.

`Key observations`:
* Refineries showed exceptional growth in both revenue and profitability.
* Trading companies appeared repeatedly among top sales, profit, and EPS growth performers.
* Finance & Investments demonstrated balanced growth across revenue, earnings, and shareholder returns.
* Some industries achieved strong revenue growth but weak profit growth, indicating margin pressure.

3. Which Companies Have the Strongest Financial Position?

Financial strength was assessed using net worth, liquidity, interest coverage, and capital structure indicators.

4. What Major Risks Were Observed?

`High Leverage Risk` : A subset of companies exhibited extremely high leverage ratios, increasing dependence on debt financing and financial risk exposure.

`Margin Compression Risk`: Certain industries showed strong sales growth but weak profit growth, suggesting rising costs and pressure on operating margins.

`Debt Concentration Risk`: Debt was heavily concentrated within banks and financial institutions. While expected due to industry structure, it remains an important risk factor that requires sector-specific interpretation.

`Growth Sustainability Risk`: Several companies reported exceptionally high growth rates, which may be influenced by low-base effects rather than sustainable long-term expansion.

* The exploratory analysis provided a comprehensive view of profitability, growth, and financial health across 4,668 listed companies. Refineries, Banking, Finance, and Technology emerged as key contributors to corporate earnings, while selected industries demonstrated strong growth momentum. Financial health analysis highlighted the importance of evaluating debt, liquidity, and leverage alongside profitability.

## KPI Creation

We have the most of the KPI already so we create only the 4 KPI's 

* `Profit Margin`
* `Debt-to-Equity-Ratio`
* `Return-on-Asset`
* `Asset-Turnover`

In [218]:
master_df['profit_margin'] = (master_df['profit_after_tax']/ master_df['sales']) * 100 # profit margin 
master_df['debt_to_equity'] = (master_df['debt'] /master_df['net_worth'])        #Debt-to-Equity-Ratio
master_df['roa'] = (master_df['profit_after_tax']/ master_df['total_assets']) * 100   # Return on Assets
master_df['asset_turnover'] = (master_df['sales']/ master_df['total_assets'])         # Asset Turnover

In [219]:
master_df[['profit_margin','debt_to_equity','roa','asset_turnover']].describe()

,profit_margin,debt_to_equity,roa,asset_turnover
count,4658.000000,4632.000000,4651.000000,4649.000000
mean,NaN,inf,-inf,0.957057
std,NaN,NaN,NaN,3.021940
min,-inf,-500.000000,-inf,-0.682202
25%,0.524126,0.005089,0.292403,0.183929
50%,5.010060,0.174342,3.222192,0.688325
75%,13.011787,0.714354,8.214335,1.163754
max,inf,inf,2895.238095,159.281739


In [220]:
(master_df['sales'] == 0).sum()

297

In [221]:
(master_df['total_assets'] == 0).sum()

2

In [222]:
(master_df['equity_capital'] == 0).sum()

1

In [223]:
# filling  the null values 
master_df['profit_margin'] = master_df['profit_margin'].replace([np.inf, -np.inf],np.nan)
master_df['roa'] = master_df['roa'].replace([np.inf, -np.inf],np.nan)
master_df['debt_to_equity'] = master_df['debt_to_equity'].replace([np.inf, -np.inf],np.nan)

In [224]:
master_df[['profit_margin','debt_to_equity','roa','asset_turnover']].describe()

,profit_margin,debt_to_equity,roa,asset_turnover
count,4367.000000,4631.000000,4649.000000,4649.000000
mean,-111.492584,0.659282,-41.629851,0.957057
std,3271.020091,10.871765,1763.965429,3.021940
min,-108050.000000,-500.000000,-85050.000000,-0.682202
25%,1.005337,0.005089,0.294654,0.183929
50%,5.315146,0.174340,3.224351,0.688325
75%,12.970916,0.714301,8.215747,1.163754
max,21605.454545,369.883212,2895.238095,159.281739


During KPI development, certain ratio calculations produced infinite values because a small number of companies had zero sales, zero assets, or zero equity. Instead of removing those companies or imputing artificial values, the infinite results were converted to missing values (NaN), preserving the integrity of the original financial data while ensuring accurate KPI analysis.

Not removing the outliers the objective of the project was corporate benchmarking across the entire listed universe. Extreme values represented genuine business conditions such as distressed companies, high-growth firms, and highly leveraged organizations. Removing them would have distorted the real financial landscape. Therefore, outliers were retained while using median-based interpretation where appropriate.

### KPI Analysis

In [225]:
# Top Profit Margin Companies
top_profit_margin = (master_df[['name','industry','profit_margin']].sort_values(by='profit_margin', ascending=False))
top_profit_margin.head(10)

,name,industry,profit_margin
420,Available Fin.,Finance & Investments,21605.454545
2500,Mehta Integrat.,Finance & Investments,15400.000000
4415,Vas Infra.,Construction,14823.809524
2078,Kaushalya Infra.,Hotels,10785.714286
3163,Quantum Digital,Packaging,9533.333333
137,Agri-Tech India,Miscellaneous,6087.500000
2370,Maha Rashtra Apx,Finance & Investments,3949.827784
3814,Softrak Venture,Finance & Investments,3600.000000
2554,MMTC,Trading,2825.842697
2803,Nuway Organic,Trading,2200.000000


In [226]:
# Top ROA Companies 
top_roa = (master_df[['name','industry','roa']].sort_values(by='roa',ascending=False))
top_roa.head(10)

,name,industry,roa
1256,Flora Textiles,Textiles - Processing,2895.238095
1901,Jagjanani Text.,Textiles - Cotton/Blended,1669.565217
4444,Ventura Textiles,Textiles - Processing,645.890411
2600,MPL Plastics,Plastics Products,348.800000
3873,Sr. Jayala. Auto,Textiles - Cotton/Blended,189.940828
3163,Quantum Digital,Packaging,125.438596
2581,Monotype India,Finance & Investments,93.832154
2217,Ksolves India,Computers - Software - Medium / Small,86.134877
1097,ECS Biztech,Computers - Software - Medium / Small,78.491620
4038,Swadeshi Polytex,Construction,73.500846


In [227]:
#Lowest  debt-to-equity
lowest_debt_to_equity = (master_df[['name','industry','debt_to_equity']].sort_values(by='debt_to_equity',ascending=True))
lowest_debt_to_equity.head(10)

,name,industry,debt_to_equity
2517,MFL India,Miscellaneous,-500.000000
1447,Golden Carpets,Textiles - Products,-93.200000
2014,Jyoti Structures,Transmisson Line Towers / Equipment,-74.974178
1135,Emami Realty Ltd,Construction,-53.712164
769,Chennai Meena,Healthcare,-34.606061
3432,Sadbhav Engg.,Construction,-29.762531
2681,Natl. Oxygen,Chemicals,-28.650000
3572,SecureKloud Tech,Computers - Software - Medium / Small,-20.081531
2503,Mena Mani Inds.,Trading,-19.937500
2377,Mahasagar Travel,Travel Agencies,-19.869565


In [228]:
#Top Asset Turnover Companies
top_asset_turnover = (master_df[['name','industry','asset_turnover']].sort_values(by='asset_turnover',ascending=False))
top_asset_turnover.head(10)

,name,industry,asset_turnover
240,Amrapali Inds.,Trading,159.281739
3111,Prithvi Exchange,Finance & Investments,70.758122
671,Burnpur Cement,Cement - North India,60.800905
3441,Sagar Diamonds,Diamond Cutting / Jewellery,39.688951
2154,KKV Agro Powers,Trading,32.451337
251,Aneri Fincap,Finance & Investments,24.500000
1171,Eraaya Lifespace,Finance & Investments,21.536232
3178,R O Jewels,Trading,17.619529
4232,Transcorp Intl.,Finance & Investments,15.865457
4636,Yuranus Infrast.,Construction,14.798387


 ## Fix outliers

In [229]:
master_df[['profit_margin','roa','debt_to_equity','asset_turnover']].quantile(0.99)

profit_margin     260.947368
roa                34.116375
debt_to_equity     11.888878
asset_turnover      5.182807
Name: 0.99, dtype: float64

In [230]:
master_df['debt_to_equity'].describe()

count    4631.000000
mean        0.659282
std        10.871765
min      -500.000000
25%         0.005089
50%         0.174340
75%         0.714301
max       369.883212
Name: debt_to_equity, dtype: float64

In [231]:
master_df['roa'].quantile(0.995)

44.02701832550165

In [232]:
master_df['profit_margin'].quantile(0.995)

1054.4995415367816

In [233]:
# profit_margin
sales_safe = master_df['sales'].replace(0, np.nan)
master_df['profit_margin'] = (master_df['profit_after_tax']/ sales_safe) * 100
# remove micro revenue companies
master_df.loc[master_df['sales'] < 100,'profit_margin'] = np.nan

In [234]:
#roa
assets_safe = master_df['total_assets'].replace(0, np.nan)
master_df['roa'] = (master_df['profit_after_tax']/ assets_safe) * 100
master_df.loc[master_df['total_assets'] < 100, 'roa'] = np.nan

In [235]:
master_df[master_df['profit_margin'] > 100][['profit_after_tax','sales','profit_margin']]

,profit_after_tax,sales,profit_margin
177,268.62,157.49,170.563210
473,7267.21,591.00,1229.646362
1878,126.02,109.66,114.918840
3055,1022.99,945.40,108.207108
3965,533.14,291.35,182.989531
4395,1240.98,357.47,347.156405
4490,235.73,206.49,114.160492


In [236]:
# Top Profit Margin Companies
top_profit_margin = (master_df[['name','industry','profit_margin']].sort_values(by='profit_margin', ascending=False))
top_profit_margin.head(10)

,name,industry,profit_margin
473,Bajaj Holdings,Finance & Investments,1229.646362
4395,Valor Estate,Construction,347.156405
3965,Sundaram Fin.Hol,Finance & Investments,182.989531
177,Alembic,Construction,170.563210
1878,IST,Auto Ancillaries,114.918840
4490,Vipul Ltd,Construction,114.160492
3055,Powergrid Infra.,Power Generation And Supply,108.207108
4113,Tata Inv.Corpn.,Finance & Investments,99.953265
413,Authum Invest,Finance & Investments,96.590623
1992,JSW Holdings,Finance & Investments,91.743336


In [237]:
master_df['profit_after_tax'].describe()

count     4664.000000
mean       323.653407
std       2774.746549
min     -31973.310000
25%          0.080000
50%          4.130000
75%         40.450000
max      69621.000000
Name: profit_after_tax, dtype: float64

In [238]:
master_df['profit_margin'].min()

-697.0163639030319

Extremely high profitability ratios were not calculation errors. They were caused by companies with extremely low revenue or asset bases, which inflated percentage metrics. To improve analytical reliability, minimum business-scale thresholds were applied before KPI interpretation.

In [239]:
final_cols = [

# identifiers
'name',
'industry',
'current_price',

# profitability
'sales',
'operating_profit',
'profit_after_tax',
'net_profit',
'opm',
'eps',
'return_on_capital_employed',
'roce3yr_avg',

# growth
'sales_growth_3years',
'sales_growth_5years',
'profit_growth_3years',
'profit_growth_5years',
'eps_growth_3years',
'eps_growth_5years',
'sales_growth_10years',
'profit_growth_10years',
'eps_growth_10years',

# financial health
'debt',
'net_worth',
'current_ratio',
'interest_coverage_ratio',
'working_capital',
'cash_equivalents',
'total_assets',
'debt_to_profit',
'leverage',
'altman_z_score',

# valuation
'market_capitalization',
'price_to_sales',
'price_to_cash_flow',
'price_to_free_cash_flow',
'peg_ratio',
'evebitda',
'market_cap_to_sales',
'enterprise_value_to_ebit',
'intrinsic_value',
'graham'
]

In [240]:
final_master = master_df[final_cols].copy()

In [241]:
print(final_master.shape)
final_master.head()

(4668, 40)


,name,industry,current_price,sales,operating_profit,profit_after_tax,net_profit,opm,eps,return_on_capital_employed,...,market_capitalization,price_to_sales,price_to_cash_flow,price_to_free_cash_flow,peg_ratio,evebitda,market_cap_to_sales,enterprise_value_to_ebit,intrinsic_value,graham
0,20 Microns,Mining / Minerals / Metals,224.55,777.49,105.65,57.38,56.16,13.59,15.89,21.70,...,792.37,1.02,14.58,24.20,0.77,7.80,1.02,8.97,246.38,189.11
1,21st Cent. Mgmt.,Finance & Investments,70.44,34.92,32.92,32.23,32.23,94.27,30.70,73.00,...,73.96,2.12,-42.02,39.91,0.05,2.25,2.12,2.25,111.73,187.38
2,360 ONE,Finance & Investments,1009.40,2920.91,1704.85,804.18,804.21,58.37,22.41,14.47,...,36627.84,12.36,-76.81,-102.75,2.72,26.30,12.36,27.20,202.65,220.16
3,3B Blackbio,Healthcare,1182.00,74.12,32.98,32.10,32.10,44.50,38.13,20.95,...,1014.49,13.55,37.34,36.64,0.72,20.72,13.55,21.21,1384.50,472.22
4,3C IT Solutions,Computers - Software - Medium / Small,44.01,61.93,3.20,1.14,1.14,5.17,31.67,25.63,...,26.50,0.44,-12.27,-10.67,NaN,9.57,0.44,10.28,NaN,NaN


final_master.to_csv('final_financial_master.csv',index=False)
print("CSV Saved")

In [242]:
check = pd.read_csv('final_financial_master.csv')
check.head()

,name,industry,current_price,sales,operating_profit,profit_after_tax,net_profit,opm,eps,return_on_capital_employed,...,market_capitalization,price_to_sales,price_to_cash_flow,price_to_free_cash_flow,peg_ratio,evebitda,market_cap_to_sales,enterprise_value_to_ebit,intrinsic_value,graham
0,20 Microns,Mining / Minerals / Metals,225.0,777.0,106.0,57.0,56.0,14.0,16.0,22.0,...,792.0,1.0,15.0,24.0,1.0,8.0,1.0,9.0,246.0,189.0
1,21st Cent. Mgmt.,Finance & Investments,70.0,35.0,33.0,32.0,32.0,94.0,31.0,73.0,...,74.0,2.0,-42.0,40.0,0.0,2.0,2.0,2.0,112.0,187.0
2,360 ONE,Finance & Investments,1009.0,2921.0,1705.0,804.0,804.0,58.0,22.0,14.0,...,36628.0,12.0,-77.0,-103.0,3.0,26.0,12.0,27.0,203.0,220.0
3,3B Blackbio,Healthcare,1182.0,74.0,33.0,32.0,32.0,44.0,38.0,21.0,...,1014.0,14.0,37.0,37.0,1.0,21.0,14.0,21.0,1384.0,472.0
4,3C IT Solutions,Computers - Software - Medium / Small,44.0,62.0,3.0,1.0,1.0,5.0,32.0,26.0,...,26.0,0.0,-12.0,-11.0,NaN,10.0,0.0,10.0,NaN,NaN


#### Keep some numeric metrics to Decimal

In [243]:
keep_decimal = [
'opm',
'return_on_capital_employed',
'roce3yr_avg',
'current_ratio',
'interest_coverage_ratio',
'debt_to_profit',
'leverage',
'altman_z_score',
'evebitda',
'enterprise_value_to_ebit',
'peg_ratio',
'price_to_sales',
'price_to_cash_flow',
'price_to_free_cash_flow',
'market_cap_to_sales'
]

In [244]:
num_cols = final_master.select_dtypes(include=['float64']).columns

In [245]:
round_cols = [c for c in num_cols if c not in keep_decimal]
final_master[round_cols] = (final_master[round_cols].round(0))

In [246]:
final_master[keep_decimal] = (final_master[keep_decimal].round(2))

In [247]:
final_master.head()

,name,industry,current_price,sales,operating_profit,profit_after_tax,net_profit,opm,eps,return_on_capital_employed,...,market_capitalization,price_to_sales,price_to_cash_flow,price_to_free_cash_flow,peg_ratio,evebitda,market_cap_to_sales,enterprise_value_to_ebit,intrinsic_value,graham
0,20 Microns,Mining / Minerals / Metals,225.0,777.0,106.0,57.0,56.0,13.59,16.0,21.70,...,792.0,1.02,14.58,24.20,0.77,7.80,1.02,8.97,246.0,189.0
1,21st Cent. Mgmt.,Finance & Investments,70.0,35.0,33.0,32.0,32.0,94.27,31.0,73.00,...,74.0,2.12,-42.02,39.91,0.05,2.25,2.12,2.25,112.0,187.0
2,360 ONE,Finance & Investments,1009.0,2921.0,1705.0,804.0,804.0,58.37,22.0,14.47,...,36628.0,12.36,-76.81,-102.75,2.72,26.30,12.36,27.20,203.0,220.0
3,3B Blackbio,Healthcare,1182.0,74.0,33.0,32.0,32.0,44.50,38.0,20.95,...,1014.0,13.55,37.34,36.64,0.72,20.72,13.55,21.21,1384.0,472.0
4,3C IT Solutions,Computers - Software - Medium / Small,44.0,62.0,3.0,1.0,1.0,5.17,32.0,25.63,...,26.0,0.44,-12.27,-10.67,NaN,9.57,0.44,10.28,NaN,NaN


In [248]:
final_master.to_csv('final_financial_master.csv',index=False)
print("CSV Updated")

CSV Updated


In [249]:
check = pd.read_csv('final_financial_master.csv')

In [250]:
check.head()

,name,industry,current_price,sales,operating_profit,profit_after_tax,net_profit,opm,eps,return_on_capital_employed,...,market_capitalization,price_to_sales,price_to_cash_flow,price_to_free_cash_flow,peg_ratio,evebitda,market_cap_to_sales,enterprise_value_to_ebit,intrinsic_value,graham
0,20 Microns,Mining / Minerals / Metals,225.0,777.0,106.0,57.0,56.0,13.59,16.0,21.70,...,792.0,1.02,14.58,24.20,0.77,7.80,1.02,8.97,246.0,189.0
1,21st Cent. Mgmt.,Finance & Investments,70.0,35.0,33.0,32.0,32.0,94.27,31.0,73.00,...,74.0,2.12,-42.02,39.91,0.05,2.25,2.12,2.25,112.0,187.0
2,360 ONE,Finance & Investments,1009.0,2921.0,1705.0,804.0,804.0,58.37,22.0,14.47,...,36628.0,12.36,-76.81,-102.75,2.72,26.30,12.36,27.20,203.0,220.0
3,3B Blackbio,Healthcare,1182.0,74.0,33.0,32.0,32.0,44.50,38.0,20.95,...,1014.0,13.55,37.34,36.64,0.72,20.72,13.55,21.21,1384.0,472.0
4,3C IT Solutions,Computers - Software - Medium / Small,44.0,62.0,3.0,1.0,1.0,5.17,32.0,25.63,...,26.0,0.44,-12.27,-10.67,NaN,9.57,0.44,10.28,NaN,NaN


In [256]:
check.dtypes

name                           object
industry                       object
current_price                 float64
sales                         float64
operating_profit              float64
profit_after_tax              float64
net_profit                    float64
opm                           float64
eps                           float64
return_on_capital_employed    float64
roce3yr_avg                   float64
sales_growth_3years           float64
sales_growth_5years           float64
profit_growth_3years          float64
profit_growth_5years          float64
eps_growth_3years             float64
eps_growth_5years             float64
sales_growth_10years          float64
profit_growth_10years         float64
eps_growth_10years            float64
debt                          float64
net_worth                     float64
current_ratio                 float64
interest_coverage_ratio       float64
working_capital               float64
cash_equivalents              float64
total_assets

## Connect with SQL 

In [251]:
pip install sqlalchemy psycopg2

Note: you may need to restart the kernel to use updated packages.


DEPRECATION: Loading egg at c:\users\vikas kumar\anaconda3\lib\site-packages\ml_project-0.0.1-py3.12.egg is deprecated. pip 24.3 will enforce this behaviour change. A possible replacement is to use pip for package installation. Discussion can be found at https://github.com/pypa/pip/issues/12330


In [252]:
from sqlalchemy import create_engine

In [253]:
engine = create_engine(
'postgresql://postgres:password077.@localhost:5432/finance_master'
)

In [259]:
df = pd.read_csv('final_financial_master.csv')
print(df.shape)

(4668, 40)


In [260]:
df.to_sql('financial_master',engine,if_exists='replace',index=False)
print("Upload Complete")

Upload Complete
